File: CodeFlow.Api/TokenTracking/TokenUsageAggregator.cs
using System.Text.Json;
using CodeFlow.Api.Dtos;
using CodeFlow.Persistence;

namespace CodeFlow.Api.TokenTracking;

/// <summary>
/// Pure in-memory rollup of <see cref="TokenUsageRecord"/> rows for a single trace. All
/// aggregations are computed on read so historical and live views share one code path. No
/// persisted rollups, no materialization — slice 1's contract.
/// </summary>
/// <remarks>
/// <para>
/// Each rollup is a flattened sum of every numeric leaf in the provider-reported
/// <c>UsageJson</c>, keyed by dotted JSON path (e.g.,
/// <c>output_tokens_details.reasoning_tokens</c>). This keeps the aggregator schema-less and
/// future-proof: when a provider adds a new field, summing it Just Works without an aggregator
/// change. The UI gets to choose which fields to render.
/// </para>
/// <para>
/// Each rollup also carries a per-(provider, model) breakdown — always populated, even when
/// there's only one combination, so the inspector doesn't have to special-case the single-combo
/// path.
/// </para>
/// </remarks>
public static class TokenUsageAggregator
{
    public static TraceTokenUsageDto Aggregate(Guid traceId, IReadOnlyList<TokenUsageRecord> records)
    {
        ArgumentNullException.ThrowIfNull(records);

        if (records.Count == 0)
        {
            return new TraceTokenUsageDto(
                TraceId: traceId,
                Total: EmptyRollup(),
                Records: Array.Empty<TokenUsageRecordDto>(),
                ByInvocation: Array.Empty<TokenUsageInvocationRollupDto>(),
                ByNode: Array.Empty<TokenUsageNodeRollupDto>(),
                ByScope: Array.Empty<TokenUsageScopeRollupDto>());
        }

        var raw = new TokenUsageRecordDto[records.Count];
        for (var i = 0; i < records.Count; i++)
        {
            raw[i] = ToRecordDto(records[i]);
        }

        return new TraceTokenUsageDto(
            TraceId: traceId,
            Total: BuildRollup(records),
            Records: raw,
            ByInvocation: BuildInvocationRollups(records),
            ByNode: BuildNodeRollups(records),
            ByScope: BuildScopeRollups(records));
    }

    private static TokenUsageRecordDto ToRecordDto(TokenUsageRecord record)
    {
        return new TokenUsageRecordDto(
            RecordId: record.Id,
            NodeId: record.NodeId,
            InvocationId: record.InvocationId,
            ScopeChain: record.ScopeChain.ToArray(),
            Provider: record.Provider,
            Model: record.Model,
            RecordedAtUtc: record.RecordedAtUtc,
            Usage: record.Usage,
            // Per-call totals are the flattened fields of this single record's usage payload.
            // Useful for the timeline view that shows each call individually.
            Totals: FlattenAndSum(new[] { record }));
    }

    private static IReadOnlyList<TokenUsageInvocationRollupDto> BuildInvocationRollups(
        IReadOnlyList<TokenUsageRecord> records)
    {
        // (NodeId, InvocationId) is the natural per-invocation key. InvocationLoop iterations
        // share the same InvocationId across rounds — actually no, each round mints a new id, so
        // this is per-LLM-round. The aggregation API conventions call this "per agent invocation"
        // but we group by InvocationId because that's the stable per-call correlator.
        return records
            .GroupBy(r => (r.NodeId, r.InvocationId))
            .Select(group => new TokenUsageInvocationRollupDto(
                NodeId: group.Key.NodeId,
                InvocationId: group.Key.InvocationId,
                Rollup: BuildRollup(group.ToArray())))
            .ToArray();
    }

    private static IReadOnlyList<TokenUsageNodeRollupDto> BuildNodeRollups(
        IReadOnlyList<TokenUsageRecord> records)
    {
        return records
            .GroupBy(r => r.NodeId)
            .Select(group => new TokenUsageNodeRollupDto(
                NodeId: group.Key,
                Rollup: BuildRollup(group.ToArray())))
            .ToArray();
    }

    private static IReadOnlyList<TokenUsageScopeRollupDto> BuildScopeRollups(
        IReadOnlyList<TokenUsageRecord> records)
    {
        // Each record's ScopeChain is `[parent, child, grandchild...]` excluding root. A scope
        // rolls up every record whose chain contains that scope id (inclusive of descendants), so
        // a parent subflow includes everything its children captured. Root totals come through
        // the trace-level rollup; we don't synthesize a "root scope" entry.
        var scopeIds = records
            .SelectMany(r => r.ScopeChain)
            .Distinct()
            .ToArray();

        return scopeIds
            .Select(scopeId => new TokenUsageScopeRollupDto(
                ScopeId: scopeId,
                Rollup: BuildRollup(records.Where(r => r.ScopeChain.Contains(scopeId)).ToArray())))
            .ToArray();
    }

    private static TokenUsageRollupDto BuildRollup(IReadOnlyList<TokenUsageRecord> records)
    {
        if (records.Count == 0)
        {
            return EmptyRollup();
        }

        var byProviderModel = records
            .GroupBy(r => (r.Provider, r.Model))
            .Select(group => new TokenUsageProviderModelTotalsDto(
                Provider: group.Key.Provider,
                Model: group.Key.Model,
                Totals: FlattenAndSum(group.ToArray())))
            .ToArray();

        return new TokenUsageRollupDto(
            CallCount: records.Count,
            Totals: FlattenAndSum(records),
            ByProviderModel: byProviderModel);
    }

    private static TokenUsageRollupDto EmptyRollup()
    {
        return new TokenUsageRollupDto(
            CallCount: 0,
            Totals: new Dictionary<string, long>(StringComparer.Ordinal),
            ByProviderModel: Array.Empty<TokenUsageProviderModelTotalsDto>());
    }

    private static IReadOnlyDictionary<string, long> FlattenAndSum(IReadOnlyList<TokenUsageRecord> records)
    {
        var totals = new Dictionary<string, long>(StringComparer.Ordinal);
        foreach (var record in records)
        {
            FlattenInto(record.Usage, prefix: string.Empty, totals);
        }
        return totals;
    }

    private static void FlattenInto(JsonElement element, string prefix, Dictionary<string, long> totals)
    {
        switch (element.ValueKind)
        {
            case JsonValueKind.Object:
                foreach (var property in element.EnumerateObject())
                {
                    var nextPrefix = prefix.Length == 0 ? property.Name : $"{prefix}.{property.Name}";
                    FlattenInto(property.Value, nextPrefix, totals);
                }
                break;

            case JsonValueKind.Number:
                if (prefix.Length == 0)
                {
                    // A bare numeric usage payload has no key — provider-shape we haven't seen.
                    // Skip silently; we never make up keys.
                    return;
                }
                if (element.TryGetInt64(out var value))
                {
                    totals[prefix] = totals.TryGetValue(prefix, out var existing)
                        ? existing + value
                        : value;
                }
                else if (element.TryGetDouble(out var doubleValue))
                {
                    // Token counts are integers in every provider we've seen, but if a provider
                    // ever reports a fractional value (cost, ratio, etc.) we round to long so the
                    // aggregator stays integer-typed. Slice 5 deliberately doesn't surface those
                    // fields in the UI; this just keeps the aggregator from crashing.
                    var rounded = (long)Math.Round(doubleValue);
                    totals[prefix] = totals.TryGetValue(prefix, out var existing)
                        ? existing + rounded
                        : rounded;
                }
                break;

            // Strings, bools, arrays, nulls are not summable. Skip silently — we don't drop them
            // from the raw `Usage` payload (the per-call DTO carries the verbatim element), but
            // they don't contribute to a sum.
            default:
                break;
        }
    }
}

File: codeflow-ui/src/app/pages/traces/token-usage-aggregator.ts
import {
  TokenUsageEventPayload,
  TokenUsageInvocationRollup,
  TokenUsageNodeRollup,
  TokenUsageProviderModelTotals,
  TokenUsageRecordDto,
  TokenUsageRollup,
  TokenUsageScopeRollup,
  TraceTokenUsageDto,
} from '../../core/models';

/**
 * Client-side mirror of `CodeFlow.Api/TokenTracking/TokenUsageAggregator.cs`. The
 * server returns one of these on initial load; for live updates we re-aggregate from
 * the same raw record list (kept in a signal) so a `TokenUsageRecorded` SSE event
 * doesn't require a server round-trip.
 *
 * Numeric leaves in the provider's `usage` object are summed by dotted JSON path
 * (e.g. `output_tokens_details.reasoning_tokens`) so a new provider field lands
 * without a code change here. Strings, booleans, arrays, and nulls are skipped from
 * sums but preserved on the per-call DTO's verbatim `usage` field.
 */
export function aggregateTokenUsage(
  traceId: string,
  records: TokenUsageRecordDto[],
): TraceTokenUsageDto {
  if (records.length === 0) {
    return {
      traceId,
      total: emptyRollup(),
      records: [],
      byInvocation: [],
      byNode: [],
      byScope: [],
    };
  }

  return {
    traceId,
    total: buildRollup(records),
    records,
    byInvocation: buildInvocationRollups(records),
    byNode: buildNodeRollups(records),
    byScope: buildScopeRollups(records),
  };
}

/** Build a per-call DTO from a `TokenUsageRecorded` SSE event payload. The server
 *  emits `recordId / nodeId / invocationId / scopeChain / provider / model / usage`
 *  but not `recordedAtUtc` or per-call totals — derive both here. */
export function recordDtoFromStreamEvent(
  event: TokenUsageEventPayload,
  recordedAtUtc: string,
): TokenUsageRecordDto {
  return {
    recordId: event.recordId,
    nodeId: event.nodeId,
    invocationId: event.invocationId,
    scopeChain: event.scopeChain ?? [],
    provider: event.provider,
    model: event.model,
    recordedAtUtc,
    usage: event.usage,
    totals: flattenAndSum([event.usage]),
  };
}

function buildInvocationRollups(records: TokenUsageRecordDto[]): TokenUsageInvocationRollup[] {
  const groups = new Map<string, TokenUsageRecordDto[]>();
  for (const record of records) {
    const key = `${record.nodeId}::${record.invocationId}`;
    const list = groups.get(key) ?? [];
    list.push(record);
    groups.set(key, list);
  }
  return Array.from(groups.entries()).map(([key, group]) => {
    const [nodeId, invocationId] = key.split('::');
    return { nodeId, invocationId, rollup: buildRollup(group) };
  });
}

function buildNodeRollups(records: TokenUsageRecordDto[]): TokenUsageNodeRollup[] {
  const groups = new Map<string, TokenUsageRecordDto[]>();
  for (const record of records) {
    const list = groups.get(record.nodeId) ?? [];
    list.push(record);
    groups.set(record.nodeId, list);
  }
  return Array.from(groups.entries()).map(([nodeId, group]) => ({
    nodeId,
    rollup: buildRollup(group),
  }));
}

function buildScopeRollups(records: TokenUsageRecordDto[]): TokenUsageScopeRollup[] {
  // Each record's ScopeChain is `[parent, child, grandchild...]` excluding root. For
  // any scope id appearing in any record's chain, sum every record whose chain
  // contains that id (inclusive of descendants).
  const scopeIds = new Set<string>();
  for (const record of records) {
    for (const id of record.scopeChain) {
      scopeIds.add(id);
    }
  }
  return Array.from(scopeIds).map(scopeId => ({
    scopeId,
    rollup: buildRollup(records.filter(r => r.scopeChain.includes(scopeId))),
  }));
}

function buildRollup(records: TokenUsageRecordDto[]): TokenUsageRollup {
  if (records.length === 0) {
    return emptyRollup();
  }

  const usageObjects = records.map(r => r.usage);

  const byProviderModel = new Map<string, TokenUsageRecordDto[]>();
  for (const record of records) {
    const key = `${record.provider}::${record.model}`;
    const list = byProviderModel.get(key) ?? [];
    list.push(record);
    byProviderModel.set(key, list);
  }

  const breakdown: TokenUsageProviderModelTotals[] = Array.from(byProviderModel.entries()).map(
    ([key, group]) => {
      const [provider, model] = key.split('::');
      return {
        provider,
        model,
        totals: flattenAndSum(group.map(r => r.usage)),
      };
    },
  );

  return {
    callCount: records.length,
    totals: flattenAndSum(usageObjects),
    byProviderModel: breakdown,
  };
}

function emptyRollup(): TokenUsageRollup {
  return {
    callCount: 0,
    totals: {},
    byProviderModel: [],
  };
}

function flattenAndSum(usageObjects: Record<string, unknown>[]): Record<string, number> {
  const totals: Record<string, number> = {};
  for (const usage of usageObjects) {
    flattenInto(usage, '', totals);
  }
  return totals;
}

function flattenInto(value: unknown, prefix: string, totals: Record<string, number>): void {
  if (value === null || value === undefined) {
    return;
  }

  if (typeof value === 'number' && Number.isFinite(value)) {
    if (prefix.length === 0) {
      // A bare numeric usage payload has no key — skip silently rather than invent one.
      return;
    }
    // Token counts are integers in every provider we've seen. If a provider ever reports
    // a fractional value, round to keep the type integer (matches the C# aggregator).
    const rounded = Math.round(value);
    totals[prefix] = (totals[prefix] ?? 0) + rounded;
    return;
  }

  if (typeof value === 'object' && !Array.isArray(value)) {
    for (const [key, child] of Object.entries(value as Record<string, unknown>)) {
      const nextPrefix = prefix.length === 0 ? key : `${prefix}.${key}`;
      flattenInto(child, nextPrefix, totals);
    }
  }

  // Strings, booleans, arrays: not summable. Skip silently — the server aggregator
  // does the same.
}
